### PR TITLE
feat(headless SSR): define the SSR Commerce Engine

### PR DESCRIPTION
**Pull request by y-lakhdar**
In this PR:
* Defining a new `SSRCommerceEngine` interface supporting two solution types (listing, search)
* Implementing a `defineCommerceEngine()` method for commerce engine configuration (to stay consistent with the non-commerce SSR package)

> Note:
> Only supporting listing page in this PR.
> The support for search page will be tackled in KIT-3394

https://coveord.atlassian.net/browse/KIT-3390

### CODE REVIEW

File: packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  buildCommerceEngine,
+} from './commerce-engine';
+
+export interface SSRCommerceEngineOptions extends CommerceEngineOptions {
+  /**
+   * The type of the solution the engine will be used for
+   * - 'search': Indicates that the engine is used for Search.
+   * - 'product-listing': Indicates that the engine is used for Product listing.
+   */
+  solutionType: 'search' | 'listing';
```
Comment (Spuffynism) [inline packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts:36]: Why do we need this? The commerce engine should work with listings, search and recs (all at the same time). Is this due to a limitation with SSR?

File: packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+/**
+ * The SSR commerce engine.
+ */
+export interface SSRCommerceEngine extends CommerceEngine {
+  /**
+   * Waits for the search to be completed and returns a promise that resolves to a `SearchCompletedAction`.
+   */
+  waitForSearchCompletedAction(): Promise<SearchCompletedAction>;
```
Comment (Spuffynism) [inline packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts:46]: `SearchCompletedAction` doesn't seem to be the right type, as it seems to be a type that works for non-commerce responses

Comment (Spuffynism) [inline packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts:46]: To better reflect that this works with listings, where we don't execute searches, I would propose we rename this
```suggestion
  waitForRequestCompletedAction(): Promise<SearchCompletedAction>;
```

**Review by fbeaudoincoveo — APPROVED** [review]

File: packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+export interface CommerceEngineDefinition<
+  TControllers extends ControllerDefinitionsMap<SSRCommerceEngine, Controller>,
+> extends EngineDefinition<
+    SSRCommerceEngine,
+    TControllers,
+    CommerceEngineOptions
+  > {}
+
+/**
+ * Initializes a Commerce engine definition in SSR with given controllers definitions and commerce engine config.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts:89]: Maybe we should clearly mark this as internal / not ready for use for now.

**Review by alexprudhomme — APPROVED** [review]

**Pull request merged**